<a href="https://colab.research.google.com/github/lenryng/AAI2025/blob/main/Code_Generation_with_ReACT_Prompting.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Exercise 2: Code Generation with ReACT Prompting

**Goal:** Generate Python code using a ReACT-style process that follows a cycle of planning, generating code, running the code, observing the output, and fixing problems when needed.

**Tools Used:**
- Google Colab
- Python
- ChatGPT

## ReACT Prompt

**Task:** Create a Python program that analyzes a list of customer order amounts.

Follow this ReACT-style process:

1. **Plan:** Briefly describe how you will solve the problem before writing the code.
2. **Act:** Generate Python code that:
   - Uses the order amounts: [120, 75, 250, 40, 180, 95, 300]
   - Calculates the total order amount
   - Calculates the average order amount
   - Finds the highest and lowest order amounts
   - Counts how many orders are above $100
   - Uses only standard Python; do not use external libraries
   - Handles an empty list without causing an error
   - Displays the results with clear labels
3. **Run:** Execute the generated Python code.
4. **Observe:** Review the output and check whether all requirements were met.
5. **Fix:** If there is an error or missing requirement, revise the code and run it again.

## ReACT Cycle

### 1. Plan

The program will store the customer order amounts in a Python list. It will calculate the total, average, highest, and lowest order amounts. It will also count the number of orders above $100. Before performing the calculations, the program will check whether the list is empty to prevent errors.

In [ ]:
# ReACT Stage 2: Act - Generate and execute the Python code

order_amounts = [120, 75, 250, 40, 180, 95, 300]

if len(order_amounts) > 0:
    total = sum(order_amounts)
    average = total / len(order_amounts)
    highest = max(order_amounts)
    lowest = min(order_amounts)
    above_100 = sum(1 for amount in order_amounts if amount > 100)

    print("Total Order Amount: $", total)
    print("Average Order Amount: $", round(average, 2))
    print("Highest Order Amount: $", highest)
    print("Lowest Order Amount: $", lowest)
    print("Number of Orders Above $100:", above_100)
else:
    print("The order list is empty. No calculations can be performed.")

Total Order Amount: $ 1060
Average Order Amount: $ 151.43
Highest Order Amount: $ 300
Lowest Order Amount: $ 40
Number of Orders Above $100: 4


### 3. Run and Observe

The code executed successfully with no errors. The output showed a total of 1,060 dollars, an average of 151.43 dollars, a highest order of 300 dollars, a lowest order of 40 dollars, and 4 orders above 100 dollars.

After reviewing the output, all required calculations are correct. However, I also need to verify that the program correctly handles an empty list as required by the prompt.

In [ ]:
# ReACT Stage 4: Test an edge case

test_orders = []

if len(test_orders) > 0:
    total = sum(test_orders)
    average = total / len(test_orders)
    highest = max(test_orders)
    lowest = min(test_orders)
    above_100 = sum(1 for amount in test_orders if amount > 100)

    print("Total Order Amount: $", total)
    print("Average Order Amount: $", round(average, 2))
    print("Highest Order Amount: $", highest)
    print("Lowest Order Amount: $", lowest)
    print("Number of Orders Above $100:", above_100)
else:
    print("The order list is empty. No calculations can be performed.")

The order list is empty. No calculations can be performed.


### 4. Fix / Reflection

The empty-list test was successful. The program displayed a clear message instead of producing an error, so no additional code fix was required.

The ReACT process helped verify the solution by following the cycle of planning the approach, generating and running the code, observing the results, and testing an edge case before accepting the final solution.